[aes,sw] Basic SW support for AES-GCM

nasahlpa opened the pull request:
This PR consists of two commits:

1. Add GCM mode in dif_aes
2. Add a minimal SW test.

The test can executed in Verilator with:
`./bazelisk.sh test --test_tag_filters=verilator --test_output=streamed //sw/device/tests:aes_gcm_test`

The ciphertext as well as the tag are compared to the expected output.

nasahlpa committed:
[aes,sw] Add DIF support for AES-GCM

This commit extends the existing dif_aes library to support
the hardware-based AES-GCM mode.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[aes,sw] Add basic AES-GCM SW test

In this commit, the previously extended dif_aes library is used
to perform a basic hardware-based AES-GCM encryption with tag
generation.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

andrea-caforio commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -300,6 +317,84 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
+  uint32_t reg = bitfield_field32_write(
+      0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, aad.valid_bytes);
+  // Put AES-GCM into AAD mode.
+  reg = bitfield_field32_write(reg, AES_CTRL_GCM_SHADOWED_PHASE_FIELD,
```
This register is written for every AAD block but it's only necessary for the first and the last block (if incomplete).

nasahlpa replied:
I hesitated doing this because it complicates the code. But I've changed it accordingly.

andrea-caforio commented on sw/device/tests/aes_gcm_test.c:
```diff
@@ -0,0 +1,128 @@
+      valid_bytes = valid_bytes - (kAesNumBlocks * sizeof(plain_text[0].data) -
+                                   sizeof(kAesModesPlainTextGcm));
+    }
+    TRY(dif_aes_set_gcm_text_mode(&aes, valid_bytes));
```
Similar to the AAD case, writing the control register is only necessary for the first and last incomplete block.

nasahlpa replied:
Changed, as discussed above.

andrea-caforio commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -81,6 +81,14 @@ typedef struct dif_aes_data {
+/**
+ * A typed representation of the AES AAD.
+ */
+typedef struct dif_aes_aad {
```
Why not add the `valid_bytes` field to the `dif_aes_data_t` struct, and then use for both AAD and message.

nasahlpa replied:
I was thinking the same. Unfortunately other tests make some assumptions on the memory layout of this struct, e.g.:
https://github.com/vogelpi/opentitan/blob/131f3a22ba17cf949167d6805d1f188f804f2784/sw/device/tests/aes_masking_off_test.c#L105-L107
Hence, I would need to change a couple of other tests and programs such that they would comply to the new struct structure. Not sure if we want to do this.

vogelpi replied:
This would be a bad idea and how it's done currently also doesn't look right to me. `dif_aes_data_t` and `dif_aes_aad_t` are structs holding one single AAD / TEXT block. But the `num_valid_bytes` just matters for the last block of both. All other blocks are 16 bytes. We don't want to have another 32-bit variable for block.

What is really needed here is a variable somewhere indicating the length of the AAD and TEXT. Then you need to compute how many blocks are required for this length and whether the last block is complete or incomplete.

nasahlpa replied:
The variable holding the length of AAD and TEXT can determined by using `sizeof(aad/text_variable)` - so I think we do not need a new variable here.

vogelpi replied:
This sounds good to me!

andrea-caforio commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -300,6 +317,84 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
+
+  // len_aad_ptx = len(aad) || len(ptx)
+  dif_aes_data_t len_aad_ptx;
+  len_aad_ptx.data[0] = __builtin_bswap32((uint32_t)(len_aad >> 32));
```
The pain of having to deal with the weird GCM encoding. :-))

vogelpi commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -235,6 +235,23 @@ dif_result_t dif_aes_start(const dif_aes_t *aes,
+      key_bytes = 32;
+    }
+
+    uint32_t reg = bitfield_field32_write(
+        0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, key_bytes);
```
What's the idea behind this? The [spec](https://docs.google.com/document/d/1CQMAL9Sgb7258iGOzDrz0KmZWY6pOvbYaAoX5iyqjFY/edit?tab=t.0) doesn't mention to configure the NUM_VALID_BYTES to the key length.

The description of the field says:
> CTRL_GCM_SHADOWED.NUM_VALID_BYTES: Indicates the number of valid bytes. Only the last block of the AAD and the plaintext/ciphertext are expected to have not all bytes valid. For all other blocks, the number of valid bytes should be set to 16.

So it should always be 16 for the initialization.

vogelpi commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -300,6 +317,87 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
   return kDifOk;
 }
 
+dif_result_t dif_aes_load_aad(const dif_aes_t *aes, const dif_aes_aad_t aad,
+                              bool set_aad_mode) {
+  if (aes == NULL) {
+    return kDifBadArg;
+  }
+
+  if (set_aad_mode) {
+    // Configure valid AAD bytes.
+    uint32_t reg = bitfield_field32_write(
+        0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, aad.valid_bytes);
+    // Put AES-GCM into AAD mode.
+    reg = bitfield_field32_write(reg, AES_CTRL_GCM_SHADOWED_PHASE_FIELD,
+                                 AES_CTRL_GCM_SHADOWED_PHASE_VALUE_GCM_AAD);
+    aes_shadowed_write(aes->base_addr, AES_CTRL_GCM_SHADOWED_REG_OFFSET, reg);
+  }
+
+  if (!aes_input_ready(aes)) {
+    return kDifUnavailable;
+  }
+
+  aes_set_multireg(aes, &aad.data[0], AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  return kDifOk;
+}
```
I would split this function into two:
- One function to configure the AAD phase (it's not a mode btw), add an argument to configure the `num_valid_bytes` field.
- Providing the data can then happen through the existing `dif_aes_load_data()` function. It needs to do exactly the same thing.

vogelpi replied:
Actually, the first fucntion can be a generic one configuring the GCM register.

nasahlpa replied:
Done, I have introduced a generic GCM register configuration function.

vogelpi replied:
Nice!

vogelpi commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -300,6 +317,87 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
+
+  // len_aad_ptx = len(aad) || len(ptx)
+  dif_aes_data_t len_aad_ptx;
+  len_aad_ptx.data[0] = __builtin_bswap32((uint32_t)(len_aad >> 32));
+  len_aad_ptx.data[1] = __builtin_bswap32((uint32_t)(len_aad & 0xFFFFFFFF));
+  len_aad_ptx.data[2] = __builtin_bswap32((uint32_t)(len_ptx >> 32));
+  len_aad_ptx.data[3] = __builtin_bswap32((uint32_t)(len_ptx & 0xFFFFFFFF));
```
You may want to check how we do this for the GCM software implementation part of cryptolib :-)

nasahlpa replied:
Yes, we do something similar [here](https://github.com/lowRISC/opentitan/blob/437e299d1060c0da4324993b9878611601ed9212/sw/device/lib/crypto/impl/aes_gcm/aes_gcm.c#L246-L248).

vogelpi commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -300,6 +317,87 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
+  return kDifOk;
+}
+
+dif_result_t dif_aes_set_gcm_tag_mode(const dif_aes_t *aes, uint64_t len_ptx,
```
Nit: this function doesn't just set the tag phase (not mode) but it actually computes the tag.

vogelpi commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -300,6 +317,87 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
+  len_aad_ptx.data[2] = __builtin_bswap32((uint32_t)(len_ptx >> 32));
+  len_aad_ptx.data[3] = __builtin_bswap32((uint32_t)(len_ptx & 0xFFFFFFFF));
+
+  aes_set_multireg(aes, &len_aad_ptx.data[0], AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  // Wait until tag generation has completed.
+  AES_WAIT_FOR_STATUS(aes, AES_STATUS_IDLE_BIT, true);
```
I would take all the data handling out. Because as is, this will only work in automatic mode. In manual mode you'll have to set the trigger which this currently does. It's probably better to have a function to configure the phase together with the number of valid bytes. And then something to drive the input data registers.

For tag, you can use a dedicated function to do the byte  swap and write the data registers. But you'll need to be able to send the trigger before reading the tag in manual mode.

vogelpi reviewed (commented):
Thanks for working on this @nasahlpa . I 've left some comments. There seems to be some confusion regarding the number of valid bytes. It's only needed for the last AAD and TEXT blocks and should be set to 16 anywhere else.

nasahlpa commented:
Thanks for the feedback @vogelpi.

I've simplified the code by introducing a generic `dif_aes_set_gcm_phase()` function that can be used to configure the phase and the valid bytes. Hence, now the already existing `dif_aes_load_data()` function can be used to load the AAD. I've kept the `dif_aes_load_gcm_tag_len()` function to have the byte swap in it.

> Thanks for working on this @nasahlpa . I 've left some comments. There seems to be some confusion regarding the number of valid bytes. It's only needed for the last AAD and TEXT blocks and should be set to 16 anywhere else.

Yes, this is what the code also did before. I just introduced the valid_bytes field in the `dif_aes_aad_t` struct to more easily keep track of the number of valid bytes.

With the new change, everywhere the `dif_aes_data_t` struct now is used. The code needs to keep track of the valid bytes in a separate variable.

vogelpi reviewed (approved):
LGTM, thanks @nasahlpa !
